### DISCUSSION TITLE
Stale Data; revalidate does not work.

### DISCUSSION THREAD

**Discussion by naishe** [vercel/next.js#51612] [Help]
### Summary
**UPDATE:** See the demo repo here: https://github.com/naishe/stale-data You can emulate the issue in `dev` mode. However, I am seeing EXACT same issue on my application in dev as well as prod (Vercel) deployments. To simplify, I haven't used auth or databases just a JSON file.

**Visual user?** See this clip I created. The top window shows the what the end user sees, and the the bottom shows the changes in `json` file which works as the database. You can see they are out of sync most of the times.

https://github.com/vercel/next.js/assets/845424/2d46ec8f-640e-4fa5-aeda-9c9bfb6eb1be

### The Setup
I have a page that displays user details (name, phone number etc) say, `/user/details/[id]`. It has an edit buton, that takes you to another page, `/user/edit/[id]`, that has a form client component.

### Data retrieval
In `/user/details/[id]`, I just pull the data from the database and render a card with the values and give a `<Link ...` to `/user/edit/${userId}`. On this page, I pull the user data again from the db, and pass it off to the  form which is a _client component_.

### The issue
The user goes to the details page sees the data in the database. Good. :) 
The user clicks edit, the form gets populated with right data. OK.
The user changes values, and the form gets submitted to `/api/users/update` using `fetch`, data gets saved, client gets 200.
On receiving 200, I redirect the user to `/user/details/${userId}` using `router.push` from `next/navigation` package (Notice this is client component, I cannot use router from `next/router`).
*I see STALE DATA on the details page.*
I click edit button, form loads, and
*I see STALE DATA in the form.*

I definitely not want to `const revalidate = 0;` or `const revalidate = 1;` on those pages, because most of the time they wouldn't want to edit the data. BUT, for experimentation, I set these values on both the pages. I wasn't sure if 0 means never validate, so I tested with 0 and 1 both, waited long enough in before clicking save to ensure the cache is expired (apparently not expired). I deployed the code on Vercel production mode. Still, I see stale data.

What am I doing wrong? 

Thanks for reading. I am open to any ideas at this moment. 

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64
    Binaries:
      Node: 20.2.0
      npm: 9.6.6
      Yarn: N/A
      pnpm: 8.5.1
    Relevant packages:
      next: 13.4.0
      eslint-config-next: 13.4.0
      react: 18.2.0
      react-dom: 18.2.0
```
```


### Example

_No response_

**Comment by nabildroid** [comment]
Same problem, seached everything and couldn't find a fix. I wonder why no one is talking about this. Revalidation is not working
RevalidatePath doesn't work
RevalidateTag doesn't work either!

**Reply by naishe** [reply]
@nabildroid -- I hope the gods of NextJS and RSC would be to listen to my plea in the heavy drum beating of React going back to PHP but failing to do that right. 😈 Here, I posted this as a desperate measure: https://twitter.com/naishe/status/1671741466086092800

**Reply by icyJoseph** [reply]
A bit hard without seeing code. You are writing what you think your code is doing. I saw this yesterday, after office hours, and it was just overwhelming. Let's see now after a fresh cup of coffee.

To be fair though, I had an issue with a page that used cookies. On one route I'd change the cookies server side, and then navigating to the cookie view page, I'd see stale data.

> - https://www.inngest.com/blog/5-lessons-learned-from-taking-next-js-app-router-to-production

The thing here is that there's a client side cache for Server Components. In my opinion this is the last sharp edge to fix up in the App Router, we are so close to merely reproduce the marvel of engineering that Pages Dir is. 

For now what I found worked for me, was to revalidate the path using server actions, not revalidate on the API. 

Here you can see: https://github.com/icyJoseph/next-js-react-gbg/blob/feat/app-router-rsc/demo-app/src/app/pokemon/capture/page.tsx

I am not such a fan of this though. In my case, the usage of cookies should make that route fully dynamic, never client side cached, but whatever 😅 - Notice that I still make an API call, I just "abuse" of the server action to bust the cache.

```ts
 const response = await fetch("/api/capture", {
        method: "POST",
        body: JSON.stringify({ id: target.id }),
        headers: { "Content-Type": "application/json" },
        signal: controller.signal,
      });

// At this point the cookie has been mutated, refresh the collection
startTransition(() => updateCollection());
```

**Reply by icyJoseph** [reply]
And regarding the smart comment about going full PHP, that's sort of the problem here, this is not doing that. This is but a Pages Dir on steroids if anything. Don't buy the media/influencers propaganda, both in favour and against.

> There's tons of media communication going like, wow full server, no more client, painting that as a genius idea or a mistake, but that's nothing but opinion leaders duking it out - let's not feed that machine

I think you do good with testing and trying out, I've build a couple of App Dir apps, and it is funny how in a way, it is just more of the same. Only now you can carve out the parts that are static, or would want to be exclusively server side renderer, or that could leverage of server API power, while keeping the good old architecture in place. Client Components should've been named, Your Good Old Components, but that's probably too long of a name.

**Reply by nabildroid** [reply]
Code to reproduce the bug
https://codesandbox.io/p/sandbox/next-js-fxis37

**Reply by icyJoseph** [reply]
There's nothing in that sandbox, but the starter code

**Reply by nabildroid** [reply]
https://codesandbox.io/p/sandbox/clever-shape-455pvc

To Reproduce
run npm run build then npm run start then visit twice /post/hello you can see the catch hit but when calling the API /api/revalidate then revisit /post/hello the data remains invalidated and the cache always hit never miss

Describe the Bug
the on-demand revalidation in production doesn't work either by calling revalidatePath not by RevalidateTag or both of them simultaneously

Expected Behavior
when calling /api/revalidate then revisiting /post/hello the page get revalidated and new data will appear, ISR after all!

**Reply by icyJoseph** [reply]
Haven't run it yet, but I think there's a small misunderstanding here. When you revalidate _inside_ an API route, you are running that revalidate server side, for all intends an purposes, on another machine. 

So then the response arrives to the client, and there's no way for the client to know that revalidate was called, nor for the server, to explicitly add that info in the response.

The revalidate applies for the "next" request. However, your client is using its cache.

> One could maybe argue that Next.js should add an (invisible to us) header in the response, etc, and the client use that to bust the client cache. It feels odd to do that though.

So you need to call `revalidatePath` so that the client knows about it, and that you do with `Server Actions`:

For example:

```tsx
const CapturePage = () => {
  async function updateCollection() {
    "use server";
    revalidatePath("/collection");
  }
  return <PokemonCapture updateCollection={updateCollection} />;
};
```

And then somewhere, after I have a response from API route:

```tsx
      const response = await fetch("/api/capture", {
        method: "POST",
        body: JSON.stringify({ id: target.id }),
        headers: { "Content-Type": "application/json" },
        signal: controller.signal,
      });

      // At this point the cookie has been mutated, refresh the collection
      startTransition(() => updateCollection());

```

**Reply by nabildroid** [reply]
yes i know, the first request after revalidation will still be a old data cache and the next requests will receive the new data
but the bug is her, all requests after revalidation are the same old cache even if you refresh the page 100 times

**Reply by nabildroid** [reply]
my usecase is a website /posts/[id] when updating the post and revalidating the path, website /post/.. will never ever reflect the new data

**Reply by nabildroid** [reply]
i'm using only server side, no client-side component

**Reply by icyJoseph** [reply]
Server Actions are available in server components anyway, but let me see your issue.

**Reply by icyJoseph** [reply]
@nabildroid  In your case the problem is a bit insidious... 

Your `/api/revalidate/route.ts`- is being statically rendered... Yah, even a route can be statically rendered, believe it or not.

```console
├ ○ /api/revalidate                        0 B                0 B

○  (Static)  automatically rendered as static HTML (uses no initial props)
```

Place a `export const dynamic = "force-dynamic";` on it.

```ts
import { revalidatePath, revalidateTag } from "next/cache";
import { NextResponse } from "next/server";

export const dynamic = "force-dynamic";

export async function GET(req: Request) {
  revalidatePath("/post/1");
  revalidateTag("hello");

  return NextResponse.json({
    message: "Hello from Reva",
    revalidated: true,
  });
}

```

And voíla. 

```console
├ λ /api/revalidate                        0 B                0 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
```

I got bitten by this on an open graph image I had done... it was always returning the same image for all my blog posts 😅 

I've should've read the docs:

- https://nextjs.org/docs/app/building-your-application/routing/router-handlers#static-route-handlers

**Reply by nabildroid** [reply]
@icyJoseph you made my day 😃 thank you so much, i been straggling with this for so long

**Reply by naishe** [reply]
@icyJoseph , @nabildroid -- you can see the updated original post for more details. I have created a repo and added a demo video.

**Reply by icyJoseph** [reply]
And you want this behaviour?


https://github.com/vercel/next.js/assets/21013447/23dee880-7b59-4ae4-942f-1007974c9fef

**Reply by naishe** [reply]
Precisely. The user always sees the data in the database. However, it would be nice (not important, at the moment), if

- the client side routing via `<Link...` still works and we don't have to use `<a ...` for redirection.
- `router.push(...)` still works and I don't do `window.location` manipulations
- The cache is not invalidated unless the data is edited. That means if I see detail page, then click edit to go to the edit page, but click on breadcrumb to go back to detail page instead of clicking submit, it does not pull from the server again.

But none of these are important for me at the moment as long as I see fresh data all the time. :)

**Reply by naishe** [reply]
@icyJoseph -- I am sorry that I didn't acknowledge the time and effort you are putting into this issue; but am glad that there is someone more experienced in NextJs than me out there to hear. I really appreciate it. 🙏

**Reply by icyJoseph** [reply]
no worries, sorry I didn't post the code, haha I got a busy, and miss-clicked enter...

I might be wrong, but one issue here is that your data doesn't come through fetch, so I think it is aggressively cached.

- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#data-fetching-without-fetch

And the fix was by no means obvious, I had to activate server actions, and pass one to the TodoForm:

```tsx
// app/TodoForm.tsx
type Props = {
  todo?: Todo;
  onSubmit: VoidFunction;
};

export function TodoForm({ todo, onSubmit }: Props) {
// rest of file
```

Added:

```tsx
// app/page.tsx
export const dynamic = "force-dynamic";
```

And:

```tsx
import { TodoDao } from "@/app/TodoDao";
import { TodoForm } from "../../TodoForm";
import { revalidatePath } from "next/cache";

type Params = {
  id: string;
};

type Props = {
  params: Params;
};

export const dynamic = "force-dynamic";

export default async function EditTodo({ params: { id } }: Props) {
  const todo = await TodoDao.getInstance().getTodo(id);

  async function onEditTodo() {
    "use server";
    revalidatePath("/");
  }

  return (
    <div>
      <h1 className="text-2xl font-bold inline">Edit Todo</h1>
      <TodoForm todo={todo} onSubmit={onEditTodo} />
    </div>
  );
}
```

As well as:

```tsx
import { revalidatePath } from "next/cache";
import { TodoForm } from "../TodoForm";

export default function AddTodo() {
  async function onAddTodo() {
    "use server";
    revalidatePath("/");
  }

  return (
    <div>
      <h1 className="text-2xl font-bold inline">Add Todo</h1>
      <TodoForm onSubmit={onAddTodo} />
    </div>
  );
}
```

That Server Action could be placed in its own file with a top level `'use server'`:

Last but not least:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    serverActions: true,
  },
};

module.exports = nextConfig;
```

I had a similar issue with a page that used cookies though, and I am not sure how I feel about having to use `Server Actions` as the fix, but maybe I am just being dense... More here: https://github.com/vercel/next.js/discussions/51488#discussioncomment-6216934

**Reply by naishe** [reply]
I will check this out. Thanks so much! I don't have really good feeling about this. It feels like band-aid solution. And, I don't like the fact that the most basic flow is not working out of the box.

**Reply by icyJoseph** [reply]
The thing is that your flow is not fetch based, and triggers a whole bunch of expected problems.

Anyway, to be fair, mutation was never fully claimed by App Router stability, it had been a  `todo` since day 1. The server actions are aimed at completing that picture.

**Comment by kevincobain2000** [comment]
Same issue link: https://github.com/vercel/next.js/issues/51788

**Comment by maulerjan** [comment]
Can we get a response from code owners please? This is a serious problem. There is currently no way for a page/route to guarantee fresh data on navigation.

**Reply by Enkratia** [reply]
Sorry for bothering, did you find alternative (rather then usage of experimental feature `staleTimes`, which still doesn't work on back/forward browser buttons + shared layouts)?

**Comment by arswarog** [comment]
I have some issue an I can't use server action - I use client side form library

**Comment by BashkirovN** [comment]
Someday this issue will be addressed. Just posting this to keep the tread alive.

**Comment by accountunavaible** [comment]
any fix?

**Comment by mehedikhan72** [comment]
any fix?
